services/streammanager/firehose/firehosemanager.go

```diff
@@ -1,159 +1,86 @@
+	client FireHoseClient
+}
 
-func init() {
-	pkgLogger = logger.NewLogger().Child("streammanager").Child("firehose")
+type FireHoseClient interface {
+	PutRecord(input *firehose.PutRecordInput) (*firehose.PutRecordOutput, error)
 }
 
 // NewProducer creates a producer based on destination config
-func NewProducer(destinationConfig interface{}, o Opts) (firehose.Firehose, error) {
-	var config Config
-	jsonConfig, err := json.Marshal(destinationConfig)
-	if err != nil {
-		return firehose.Firehose{}, fmt.Errorf("[FireHose] Error while marshalling destination config :: %w", err)
-	}
-	err = json.Unmarshal(jsonConfig, &config)
+func NewProducer(destinationConfig interface{}, o common.Opts) (*FireHoseProducer, error) {
+	sessionConfig, err := awsutils.NewSessionConfig(destinationConfig, o.Timeout, firehose.ServiceName)
 	if err != nil {
-		return firehose.Firehose{}, fmt.Errorf("[FireHose] error  :: error in firehose while unmarshelling destination config:: %w", err)
-	}
-	var s *session.Session
-	httpClient := &http.Client{
-		Timeout: o.Timeout,
-	}
-
-	if config.AccessKeyID == "" || config.AccessKey == "" {
-		s = session.Must(session.NewSession(&aws.Config{
-			Region:     aws.String(config.Region),
-			HTTPClient: httpClient,
-		}))
-	} else {
-		s = session.Must(session.NewSession(&aws.Config{
-			HTTPClient:  httpClient,
-			Region:      aws.String(config.Region),
-			Credentials: credentials.NewStaticCredentials(config.AccessKeyID, config.AccessKey, ""),
-		}))
+		return nil, err
 	}
-	var fh *firehose.Firehose = firehose.New(s)
-	return *fh, nil
+	return &FireHoseProducer{client: firehose.New(awsutils.CreateSession(sessionConfig))}, nil
 }
 
 // Produce creates a producer and send data to Firehose.
-func Produce(jsonData json.RawMessage, producer, destConfig interface{}) (statusCode int, respStatus, responseMessage string) {
+func (producer *FireHoseProducer) Produce(jsonData json.RawMessage, destConfig interface{}) (int, string, string) {
 	parsedJSON := gjson.ParseBytes(jsonData)
-	var putOutput *firehose.PutRecordOutput
-	var errorRec error
-
-	fh, ok := producer.(firehose.Firehose)
-
-	if !ok {
-		respStatus = "Failure"
-		responseMessage = "[FireHose] error :: Could not create producer"
-		return 400, respStatus, responseMessage
-	}
-	var config Config
-	jsonConfig, err := json.Marshal(destConfig)
-	if err != nil {
-		respStatus = "Failure"
-		responseMessage = "[FireHose] error :: " + err.Error()
-		pkgLogger.Errorf("[FireHose] error  :: %v", err)
-		statusCode := 400
-		return statusCode, respStatus, responseMessage
-	}
-	err = json.Unmarshal(jsonConfig, &config)
-	if err != nil {
-		respStatus = "Failure"
-		responseMessage = "[FireHose] error  :: " + err.Error()
-		pkgLogger.Errorf("[FireHose] error  :: %v", err)
-		statusCode := 400
-		return statusCode, respStatus, responseMessage
+	client := producer.client
+	if client == nil {
+		return 400, "Failure", "[FireHose] error :: Could not create producer"
```

Comment:
Wouldn't it be better to validate this at the "NewProducer" stage?

Reply:
We can but it is not necessary that we only use NewProducer to create the producer right, especially for testing purposes we pass our own mock client so to catch that this check is added.

Reply:
Makes sense.